### ISSUE TITLE
HelmRelease.helm.toolkit.fluxcd.io is invalid: spec.chart: Required value -   when using spec.chartref

### ISSUE THREAD

**Issue by UriZafrir** [fluxcd/flux2#5116]
### Describe the bug

Hi all,
I am trying to install spark from oci:
https://artifacthub.io/packages/helm/bitnami/spark
```
apiVersion: v1
kind: Namespace
metadata:
  name: spark
---
apiVersion: source.toolkit.fluxcd.io/v1beta2
kind: OCIRepository
metadata:
  name: bitnami-spark
  namespace: flux-system
spec:
  interval: 10m
  url: oci://registry-1.docker.io/bitnamicharts/spark
---
apiVersion: helm.toolkit.fluxcd.io/v2beta1
kind: HelmRelease
metadata:
  name: spark
  namespace: spark
spec:
  interval: 5m
  releaseName: spark
  chartRef:
    kind: OCIRepository
    name: bitnami-spark
    namespace: flux-system
  values:
```

but getting the following error:

> Detecting drift for revision dev@sha1:db3ca47946e8fad25dfed4d00ae7278f568abeee with a timeout of 9m30sHelmRelease/spark/spark dry-run failed (Invalid): HelmRelease.helm.toolkit.fluxcd.io "spark" is invalid: spec.chart: Required value

Which is pretty strange since in the docs it says never to use both spec.chart and spec.chartRef together..
https://fluxcd.io/flux/components/helm/helmreleases/#chart-reference
![image](https://github.com/user-attachments/assets/d9d2154a-24f9-46d7-9c82-588355fcae60)

I must be missing something here..
Would love for assistance!

### Steps to reproduce

install flux,
apply manifests

### Expected behavior

installed 

### OS / Distro

Ubuntu 22.04

### Flux version

flux: v2.4.0

### Flux check

 flux check
► checking prerequisites
✔ Kubernetes 1.30.6+k3s1 >=1.28.0-0
► checking version in cluster
✔ distribution: flux-v2.4.0
✔ bootstrapped: true
► checking controllers
✔ helm-controller: deployment ready
► ghcr.io/fluxcd/helm-controller:v1.1.0
✔ kustomize-controller: deployment ready
► ghcr.io/fluxcd/kustomize-controller:v1.4.0
✔ notification-controller: deployment ready
► ghcr.io/fluxcd/notification-controller:v1.4.0
✔ source-controller: deployment ready
► ghcr.io/fluxcd/source-controller:v1.4.1
► checking crds
✔ alerts.notification.toolkit.fluxcd.io/v1beta3
✔ buckets.source.toolkit.fluxcd.io/v1
✔ gitrepositories.source.toolkit.fluxcd.io/v1
✔ helmcharts.source.toolkit.fluxcd.io/v1
✔ helmreleases.helm.toolkit.fluxcd.io/v2
✔ helmrepositories.source.toolkit.fluxcd.io/v1
✔ kustomizations.kustomize.toolkit.fluxcd.io/v1
✔ ocirepositories.source.toolkit.fluxcd.io/v1beta2
✔ providers.notification.toolkit.fluxcd.io/v1beta3
✔ receivers.notification.toolkit.fluxcd.io/v1
✔ all checks passed

### Git provider

github

### Container Registry provider

docker

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by matheuscscp** [comment]
`apiVersion: helm.toolkit.fluxcd.io/v2beta1` -> `apiVersion: helm.toolkit.fluxcd.io/v2`

**Comment by dandrzejewski** [comment]
I believe I'm experiencing this as well, even with /v2.

here's the yaml:

```yaml
apiVersion: helm.toolkit.fluxcd.io/v2
kind: HelmRelease
metadata:
  name: keycloak
  namespace: keycloak
spec:
  interval: 5m
  timeout: 5m
  releaseName: keycloak
  chartRef:
    kind: OCIRepository
    name: keycloak
    namespace: keycloak
  values:
    # https://github.com/bitnami/charts/blob/main/bitnami/keycloak/values.yaml
    ingress:
      enabled: true
      hostnameStrict: true
```

And here's the message:

```
 HelmRelease/keycloak/keycloak dry-run failed (Invalid): HelmRelease.helm.toolkit.fluxcd.io "keycloak" is invalid: [spec.chart.spec.chart: Required value, spec.chart.spec.sourceRef: Required value, <nil>: Invalid value: "null": some validation rules were not checked because the object was invalid; correct the existing errors to complete validation]
```

**Comment by dandrzejewski** [comment]
I believe I'm experiencing the same issue, even with `/v2`.  Here's the yaml:

```yaml
apiVersion: source.toolkit.fluxcd.io/v1beta2
kind: OCIRepository
metadata:
  name: keycloak
  namespace: keycloak
spec:
  interval: 5m
  url: oci://registry-1.docker.io/bitnamicharts/keycloak
---
apiVersion: helm.toolkit.fluxcd.io/v2
kind: HelmRelease
metadata:
  name: keycloak
  namespace: keycloak
spec:
  interval: 5m
  timeout: 5m
  releaseName: keycloak
  chartRef:
    kind: OCIRepository
    name: keycloak
    namespace: keycloak
  values:
    # https://github.com/bitnami/charts/blob/main/bitnami/keycloak/values.yaml
    ingress:
      enabled: true
      hostnameStrict: true
```

And the error from the log:

> 2025-01-01T22:05:39.736Z error Kustomization/apps.flux-system - Reconciliation failed after 36.486545ms, next try in 10m0s HelmRelease/keycloak/keycloak dry-run failed (Invalid): HelmRelease.helm.toolkit.fluxcd.io "keycloak" is invalid: [spec.chart.spec.chart: Required value, spec.chart.spec.sourceRef: Required value, <nil>: Invalid value: "null": some validation rules were not checked because the object was invalid; correct the existing errors to complete validation]

**Comment by stefanprodan** [comment]
Migrating a HelmRelease from v2beta to v2 and from `chart` to `chartRef` can't be done in a single commit unless you're using [Flux Operator](https://github.com/controlplaneio-fluxcd/flux-operator) who knows how to upgrade the etcd storage when a Flux CRD gains a new version.

The procedure to move to OCIRepository requires 2 commits:
1. In Git set `apiVersion: helm.toolkit.fluxcd.io/v2` and add an annotation to the HelmRelease e.g. `fluxcd.io/version: v2`.
2. Commit, push changes and wait for Flux to reconcile (at this point the etcd storage has been updated to v2).
3. Switch from `chart` to `chartRef` and push the changes to the repo (the update should work now).

**Comment by stefanprodan** [comment]
@frank-bee can you post here the output of `flux check`

**Comment by frank-bee** [comment]
➜  gitops git:(main) ✗ flux check                                                                                              

► checking prerequisites
✔ Kubernetes 1.31.6 >=1.30.0-0
► checking version in cluster
✔ distribution: flux-v2.5.1
✔ bootstrapped: true
► checking controllers
✔ helm-controller: deployment ready
► ghcr.io/fluxcd/helm-controller:v1.2.0
✔ kustomize-controller: deployment ready
► ghcr.io/fluxcd/kustomize-controller:v1.5.1
✔ notification-controller: deployment ready
► ghcr.io/fluxcd/notification-controller:v1.5.0
✔ source-controller: deployment ready
► ghcr.io/fluxcd/source-controller:v1.5.0
► checking crds
✔ alerts.notification.toolkit.fluxcd.io/v1beta3
✔ buckets.source.toolkit.fluxcd.io/v1
✔ gitrepositories.source.toolkit.fluxcd.io/v1
✔ helmcharts.source.toolkit.fluxcd.io/v1
✔ helmreleases.helm.toolkit.fluxcd.io/v2
✔ helmrepositories.source.toolkit.fluxcd.io/v1
✔ kustomizations.kustomize.toolkit.fluxcd.io/v1
✔ ocirepositories.source.toolkit.fluxcd.io/v1beta2
✔ providers.notification.toolkit.fluxcd.io/v1beta3
✔ receivers.notification.toolkit.fluxcd.io/v1
✔ all checks passed

**Comment by frank-bee** [comment]
@stefanprodan sure...

```
(☸|aks-cluster-copperdot-ecs-dev:N/A)➜  gitops git:(main) ✗ kubectl get kustomization apps -n flux-system -o yaml
apiVersion: kustomize.toolkit.fluxcd.io/v1
kind: Kustomization
metadata:
  creationTimestamp: "2025-05-12T13:55:14Z"
  finalizers:
  - finalizers.fluxcd.io
  generation: 1
  labels:
    kustomize.toolkit.fluxcd.io/name: flux-system
    kustomize.toolkit.fluxcd.io/namespace: flux-system
  name: apps
  namespace: flux-system
  resourceVersion: "230876390"
  uid: 46bb622b-d5e8-4225-9015-24b5fdbae054
spec:
  dependsOn:
  - name: configs
  force: false
  interval: 1m0s
  path: ./apps/dev
  postBuild:
    substitute:
      environment: dev
  prune: true
  sourceRef:
    kind: GitRepository
    name: flux-system
status:
  conditions:
  - lastTransitionTime: "2025-05-12T17:21:51Z"
    message: Detecting drift for revision main@sha1:daa64a82ad58bdcd30f51ad9724ddb2fd9a0be68
      with a timeout of 30s
    observedGeneration: 1
    reason: ProgressingWithRetry
    status: "True"
    type: Reconciling
  - lastTransitionTime: "2025-05-12T17:21:51Z"
    message: |
      HelmRelease/tdp/tdp-api dry-run failed (Invalid): HelmRelease.helm.toolkit.fluxcd.io "tdp-api" is invalid: [spec.chart.spec.chart: Required value, spec.chart.spec.sourceRef: Required value, <nil>: Invalid value: "null": some validation rules were not checked because the object was invalid; correct the existing errors to complete validation]
    observedGeneration: 1
    reason: ReconciliationFailed
    status: "False"
    type: Ready
  lastAttemptedRevision: main@sha1:daa64a82ad58bdcd30f51ad9724ddb2fd9a0be68
  observedGeneration: -1

```

**Comment by stefanprodan** [comment]
Ok so the issue is indeed with the CRD, run `flux uninstall --keep-namespace` then do `kubectl get crds` and check if the helmrelease CRD is gone. Rerun bootstrap with flux CLI v2.5.1 to restore the cluster.

**Comment by frank-bee** [comment]
@stefanprodan I did this already, but let me try again...

**Comment by frank-bee** [comment]
@stefanprodan I uninstalled flux and checked if there are CRDs:
```
(☸|aks-cluster-copperdot-ecs-dev:N/A)➜  ~ k get crds | grep flux 
```
There were not CRDs left.

After re-bootstrapping I get the same issue again.

Is it possible that some flux resources will older versions are "stuck in etcd" ?
How can I force the flux controllers to ignore those or maybe clean them up?
Unfortunately I cannot re-create the cluster easily. Accessing etcd in azure might me possible but no clue how.

**Comment by stefanprodan** [comment]
Do you have multiple Flux instances deployed on that cluster, is the Azure Flux addon enabled?

**Comment by frank-bee** [comment]
```
(☸|aks-cluster-copperdot-ecs-dev:N/A)➜  ~ az k8s-configuration flux list -g rg-aks-copperdot-ecs-dev -c aks-cluster-copperdot-ecs-dev -t managedClusters            
[]
(☸|aks-cluster-copperdot-ecs-dev:N/A)➜  ~ az k8s-extension list -g rg-aks-copperdot-ecs-dev -c aks-cluster-copperdot-ecs-dev -t managedClusters                     
[]
```

**Comment by stefanprodan** [comment]
Run `flux export hr` and add it to your Git repo and see if it errors out

**Comment by frank-bee** [comment]
@stefanprodan yes it does

<img width="681" alt="Image" src="https://github.com/user-attachments/assets/65a68578-5bab-43d2-8ea4-cb759e8cf820" />

**Comment by stefanprodan** [comment]
Ok can you run `kubectl apply --server-side --dry-run=server --field-manager=kustomize-controller -f hr.yaml` and see if it errors out the same.

**Comment by stefanprodan** [comment]
I can't explain what's going on, is like kustomize-controller talks to a different API than the CLI...

**Comment by frank-bee** [comment]
it doesn't 

```
(☸|aks-cluster-copperdot-ecs-dev:N/A)➜  gitops git:(main) ✗ kubectl apply --server-side --dry-run=server --field-manager=kustomize-controller -f f.yaml  
helmrelease.helm.toolkit.fluxcd.io/tdp-api serverside-applied (server dry run)
(☸|aks-cluster-copperdot-ecs-dev:N/A)➜  gitops git:(main) ✗ kubectl get kustomization -A                                                                
NAMESPACE     NAME          AGE    READY   STATUS
flux-system   apps          119m   False   HelmRelease/tdp/tdp-api dry-run failed (Invalid): HelmRelease.helm.toolkit.fluxcd.io "tdp-api" is invalid: [spec.chart.spec.chart: Required value, spec.chart.spec.sourceRef: Required value, <nil>: Invalid value: "null": some validation rules were not checked because the object was invalid; correct the existing errors to complete validation]...
flux-system   configs       119m   True    Applied revision: main@sha1:05c81f6d2790d67744e8d2b40d40519354f67f81
flux-system   controllers   119m   True    Applied revision: main@sha1:05c81f6d2790d67744e8d2b40d40519354f67f81
flux-system   flux-system   119m   True    Applied revision: main@sha1:05c81f6d2790d67744e8d2b40d40519354f67f81
(☸|aks-cluster-copperdot-ecs-dev:N/A)➜  gitops git:(main) ✗ 
```

I think there is a leftover of a flux resource in etcd. 
Just wonder why kustomize controller is affected by that (and  kubectl not )

**Comment by stefanprodan** [comment]
kustomize-controller uses the controller-runtime cache, assigning a service account to the HelmRelease will force kustomize-controller into asking for a fresh copy of the HelmRelease CRD from the API.

You can try setting `serviceAccountName: helm-controller` in the apps Kustomization and see if anything changes.

**Comment by frank-bee** [comment]
I tried that as well. Also removed all other versions than v2 from the CRD.
No progress.

**Comment by stefanprodan** [comment]
Looks like you need to scrap this cluster, your theory with older versions "stuck in etcd" seems valid.

**Comment by frank-bee** [comment]
@stefanprodan FYI
I have the same problem in our prod cluster. 
So this seems to be a general problem. maybe is has to do with the azure extension for flux which was installed in both cluster before I switched to manually installing flux

**Comment by frank-bee** [comment]
@stefanprodan 
As a workaround I tried this
```
---
apiVersion: source.toolkit.fluxcd.io/v1
kind: HelmRepository
metadata:
  name: azure-dev
  namespace: the-technical-details-parser
spec:
  type: "oci"
  interval: 5m0s
  provider: azure
  url: oci://foo.azurecr.io/helm-charts/
---
apiVersion: helm.toolkit.fluxcd.io/v2
kind: HelmRelease
metadata:
  name: tdp-api
  namespace: the-technical-details-parser
spec:
  chart:
    spec:
      chart: pcb-technical-details-parser-api
      version: '0.1.0-655718e'
      sourceRef:
        kind: HelmRepository
        name: azure-dev
      interval: 5m
  interval: 1m0s
```

The issue is gone now.
But there is a new one!!

`version` in the helm chart template seems to be ignored 

```
(☸|aks-cluster-copperdot-ecs-dev:N/A)➜  ~ k describe helmrelease tdp-api -n the-technical-details-parser 

Name:         tdp-api
Namespace:    the-technical-details-parser
Labels:       kustomize.toolkit.fluxcd.io/name=theapps
              kustomize.toolkit.fluxcd.io/namespace=flux-system
Annotations:  <none>
API Version:  helm.toolkit.fluxcd.io/v2
Kind:         HelmRelease
Metadata:
  Creation Timestamp:  2025-05-13T08:37:53Z
  Finalizers:
    finalizers.fluxcd.io
  Generation:        1
  Resource Version:  231279771
  UID:               90dbd762-c446-4473-8aa2-4b398ab0b26b
Spec:
  Chart:
    Spec:
      Chart:               pcb-technical-details-parser-api
      Interval:            5m
      Reconcile Strategy:  ChartVersion
      Source Ref:
        Kind:   HelmRepository
        Name:   azure-dev
      Version:  >=1.0.0-alpha
  Interval:     1m0s
  Test:
    Enable:  false
  Values:
    Ingress:
      Hosts:
        Host:  test.dev.copperdot.tools
        Paths:
          Path:       /
          Path Type:  ImplementationSpecific
Status:
  Conditions:
    Last Transition Time:  2025-05-13T08:37:54Z
    Message:               Fulfilling prerequisites
    Observed Generation:   1
    Reason:                Progressing
    Status:                True
    Type:                  Reconciling
    Last Transition Time:  2025-05-13T08:38:57Z
    Message:               HelmChart 'the-technical-details-parser/the-technical-details-parser-tdp-api' is not ready: chart pull error: failed to get chart version for remote reference: could not locate a version matching provided version string >=1.0.0-alpha
    Observed Generation:   1
    Reason:                SourceNotReady
    Status:                False
    Type:                  Ready
  Helm Chart:              the-technical-details-parser/the-technical-details-parser-tdp-api
  Observed Generation:     -1
Events:
  Type    Reason            Age   From             Message
  ----    ------            ----  ----             -------
  Normal  HelmChartCreated  3m4s  helm-controller  Created HelmChart/the-technical-details-parser/the-technical-details-parser-tdp-api with SourceRef 'HelmRepository/the-technical-details-parser/azure-dev'

```

**Comment by frank-bee** [comment]
@stefanprodan in a fresh cluster this issue was gone, you might close the issue

**Comment by frank-bee** [comment]
thanks a lot for you help!
